File: leetcode-js/309-best-time-to-buy-and-sell-stock-with-cooldown.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    let profit1 = 0;
    let profit2 = 0;

    for (let i = 1; i < prices.length; i += 1) {
        const tmp = profit1;
        profit1 = Math.max(profit1 + prices[i] - prices[i-1], profit2);
        profit2 = Math.max(profit2, tmp);
    }

    return Math.max(profit1, profit2);
};

console.log(maxProfit([1,2,3,0,2]));

File: leetcode-java/309-best-time-to-buy-and-sell-stock-with-cooldown.java
public class Solution {
    public int maxProfit(int[] prices) {
        if (prices.length == 0) {
            return 0;
        }

        int sell = 0;
        int prevSell = 0;
        int buy = -prices[0];
        int prevBuy = 0;

        for (int price : prices) {
            prevBuy = buy;
            buy = Math.max(prevBuy, prevSell - price);
            prevSell = sell;
            sell = Math.max(prevSell, prevBuy + price);
        }

        return sell;
    }
}
